module/spl/spl-kmem.c

```diff
@@ -133,6 +133,32 @@ strfree(char *str)
 }
 EXPORT_SYMBOL(strfree);
 
+#ifndef __GFP_REPEAT
+#define	__GFP_REPEAT	__GFP_RETRY_MAYFAIL
+#endif
+void *
+spl_kvmalloc(size_t size, gfp_t flags)
+{
+	gfp_t kmalloc_flags = flags;
+	void *ret;
+#ifdef HAVE_KVMALLOC
+	if ((flags & GFP_KERNEL) == GFP_KERNEL) {
+		return (kvmalloc(size, flags));
+	}
+#endif
+	if (size > PAGE_SIZE) {
+		kmalloc_flags |= __GFP_NOWARN;
+		if (!(kmalloc_flags & __GFP_REPEAT) ||
+		    (size <= PAGE_SIZE << PAGE_ALLOC_COSTLY_ORDER)) {
```

Comment:
Good question... @BrainSlayer?

Reply:
seem to be kernel version specific. hint: https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/mm/util.c?h=next-20190719&id=cc965a29db172c28e25b9742db86a85766a08bf5

so this hack was used before __GFP_RETRY_MAYFAIL was introduced. so i assume we should use this hack if __GFP_RETRY_MAYFAIL is not available

Reply:
I just realized that spl_kvmalloc is more or less a modified copy of `kvmalloc_node` (from `mm/util.c`). Can't we just call `kvmalloc_node` if it is available and fall back to `spl_kvmalloc` if not? This way we would not need to do any kernel version checks for __GFP_RETRY_MAYFAIL...

I guess something like this could work:

~~~c
void * spl_kvmalloc(size_t size, gfp_t lflags)
{
...
#ifdef HAVE_KVMALLOC
	if ((lflags & GFP_KERNEL) == GFP_KERNEL)
		return (kvmalloc(size, lflags));
#elifdef HAVE_KVMALLOC_NODE
	return kmalloc_node(size, lflags, NUMA_NO_NODE);
#else
	/* current spl_kvmalloc implementation here */
#endif
}
~~~

Reply:
> Can't we just call kvmalloc_node if it is available and fall back to spl_kvmalloc if not?

We can, but let's go one step farther to keep this simple.  Since only the 4.12 kernel requires the hack and it's not a long term kernel let's either use `kmalloc_node` or just fallback to our own implementation.  Something like this (untested).  The autoconf check will need to be updated of course to check for `kvmalloc_node()` and set `HAVE_KVMALLOC_NODE` instead.

```c
void *
spl_kvmalloc(size_t size, gfp_t lflags)
{
#if defined(HAVE_KVMALLOC_NODE) && defined(__GFP_RETRY_MAYFAIL)
	/*
	 * For 4.13 and newer kernels which provide kvmalloc_node() and the
	 * __GFP_RETRY_MAYFAIL flag kvmalloc_node() can be used.
	 */
	return (kvmalloc_node(size, lflags, NUMA_NO_NODE));
#else
	gfp_t kmalloc_lflags = lflags;

	if (size > PAGE_SIZE) {
		/*
		 * We need to set __GFP_NOWARN here since spl_kvmalloc is not
		 * only called by spl_kmem_alloc_impl but can be called
		 * directly with custom lflags, too. In that case
		 * kmem_flags_convert does not get called, which would
		 * implicitely set __GFP_NOWARN.
		 */
		kmalloc_lflags |= __GFP_NOWARN;

		/*
		 * We have to override __GFP_RETRY_MAYFAIL by __GFP_NORETRY
		 * for !costly requests because there is no other way to tell
		 * the allocator that we want to fail rather than retry
		 * endlessly.
		 */
		if (!(kmalloc_lflags & __GFP_REPEAT) ||
		    (size <= PAGE_SIZE << PAGE_ALLOC_COSTLY_ORDER))
			kmalloc_lflags |= __GFP_NORETRY
	}

	void *ptr = kmalloc_node(size, kmalloc_lflags, NUMA_NO_NODE);
	if (ptr != NULL || size <= PAGE_SIZE)
		return (ptr);

	return (__vmalloc(size, lflags | __GFP_HIGHMEM, PAGE_KERNEL));
#endif
}
```

Reply:
> Since only the 4.12 kernel requires the hack

No, any kernel <4.12 does require the hack.

Reply:
Sorry, I meant we use our own implementation which includes the same hack for any kernel which does not provide `__GFP_RETRY_MAYFAIL`, so <= 4.12.  See the proposed updated function.  Note, this implementation depends on us **not** defining `__GFP_RETRY_MAYFAIL` for older kernels.

Reply:
Ok, that makes sense. What about @brainslayer's objection?

"kvmalloc is a stable api. kvmalloc_node is not and may change. so using kvmalloc is more safe for forward compatibility".

> Note, this implementation depends on us not defining __GFP_RETRY_MAYFAIL for older kernel
>[edit] Alternately we could include the check for if __GFP_RETRY_MAYFAIL is defined in the configure check.

Yep, I'll adapt this as soon as we have agreed on one of the variants

Reply:
> kvmalloc is a stable api

That's a good point, and I'm all for preventing future compatibility issues.  Calling `kvmalloc()` makes sense to me if the autoconf check also determines that `__GFP_RETRY_MAYFAIL` is defined (so we have the 4.13+ version).

Reply:
Hm. This feels like we're making things more complicated than we should... when available, `kvmalloc` in any case does the right thing for us...
- 4.12 has `kvmalloc` including the hack
- anything after that has `kvmalloc` without the hack
- anything <4.12 needs our own implementation anyways

We should drop this idea of using kvmalloc_node as IMHO there is no benefit at all.

Reply:
but what if i say that some kernels including 4.4 and 4.9 have kvmalloc_node, but not kvmalloc support. but again with a different implementation than our fallback method. but personally think that our fallback method is better for 4.4 and 4.9 etc. kernels

Reply:
@Brainslayer I can't find `kvmalloc_node` in 4.4 or 4.9 o.O

~~~
c0d3@z3r0:[~/src/linux.git]# git grep kvmalloc_node up/stable/linux-4.9.y
c0d3@z3r0:[~/src/linux.git]# git grep kvmalloc_node up/stable/linux-4.4.y
c0d3@z3r0:[~/src/linux.git]# 
~~~

Reply:
okay. could be a private kernel version by me then. its not all vanilla on my systems

Reply:
There's one small additional concern.  We can't quite use `kvmalloc` unconditionally until 4.18 when this commit was merged, torvalds/linux@ce91f6ee5b3bbbad8caff61b1c46d845c8db19bf.  We'll need to do something similar in the SPL code.

> but what if i say that some kernels including 4.4 and 4.9 have kvmalloc_node, but not kvmalloc support

We should check for `kvmalloc` support.  And currently, it looks like neither 4.9.186 or 4.4.186 has backported this functionality.  And if they ever do I'd hope they backport all of it.

Reply:
> There's one small additional concern. We can't quite use `kvmalloc` unconditionally until 4.18 when this commit was merged, [torvalds/linux@ce91f6e](https://github.com/torvalds/linux/commit/ce91f6ee5b3bbbad8caff61b1c46d845c8db19bf). We'll need to do something similar in the SPL code.

Do you mean that fallback to `kmalloc_node` for !GPF_KERNEL? We already do this, don't we?

~~~c
#ifdef HAVE_KVMALLOC
...
	if ((lflags & GFP_KERNEL) == GFP_KERNEL)
		return (kvmalloc(size, lflags));
#endif

...
	/*
	 * For non-GFP_KERNEL allocations we stick to kmalloc_node
	 * and __vmalloc depeding on the allocation size.
	 */
	ptr = kmalloc_node(size, kmalloc_lflags, NUMA_NO_NODE);
	if (ptr || size <= PAGE_SIZE) {
		return (ptr);
	} else {
		return (__vmalloc(size, lflags | __GFP_HIGHMEM,
		    PAGE_KERNEL));
	}
~~~

Reply:
Mostly, but the fallback logic doesn't appear to entirely handle this case.  Specifically, we don't want to end up calling `__vmalloc` for `!GFP_KERNEL` since it internally performs `GFP_KERNEL` allocations.

Reply:
it should be considered that i never intendet to use kvmalloc for GFP_ATOMIC style allocations. its no "one shit for all" allocator. this is why i originally only implemented it for safe locations. but in any case where GFP_KERNEL isnt defined, the first try is kmalloc_node. if this fails, we may return NULL here in  !GFP_KERNEL case if you feel more safe with it. 
so if (ptr || size <= PAGE_SIZE || !(lflags & GFP_KERNEL))

Reply:
but i'm also not sure if this is a big problem here. the chance for having a wait problem with __vmalloc even if GFP_KERNEL is used internally is low or non existing for most cases. its the question how its used. for DMA transactions we cannot use anyway. and GFP_ATOMIC is only required for time critical contexts like irq's etc. so most zfs code locations are safe to use the way we are going here.

Reply:
> Mostly, but the fallback logic doesn't appear to entirely handle this case. Specifically, we don't want to end up calling `__vmalloc` for `!GFP_KERNEL` since it internally performs `GFP_KERNEL` allocations.

> but in any case where GFP_KERNEL isnt defined, the first try is kmalloc_node. if this fails, we may return NULL here in !GFP_KERNEL case if you feel more safe with it.
> so if (ptr || size <= PAGE_SIZE || !(lflags & GFP_KERNEL))

what about this?

~~~c
	/*
	 * For non-GFP_KERNEL allocations we stick to kmalloc_node
	 * and __vmalloc depeding on the allocation size.
	 */
	if (size <= PAGE_SIZE || (lflags & GFP_KERNEL) != GFP_KERNEL) {
		ptr = kmalloc_node(size, kmalloc_lflags, NUMA_NO_NODE);
		if (ptr)
			return (ptr);

		/*
		 * We cannot fall back to __vmalloc for non-GFP_KERNEL
		 * allocations when kmalloc failed because __vmalloc
		 * internally uses GPF_KERNEL allocations. Stop here.
		 */
		if ((lflags & GFP_KERNEL) != GFP_KERNEL))
			return NULL;
	}

	return (__vmalloc(size, lflags | __GFP_HIGHMEM, PAGE_KERNEL));
~~~

Reply:
wrong behaviour. if kmalloc_node fails and size <= PAGE_SIZE must return NULL too. my small line mod works perfect here. it calls kmalloc_node, no matter which flags and returns NULL if allocations fails and size <= PAGE_SIZE or flags !GFP_KERNEL

Reply:
in your code if size is > PAGE_SIZE and flags are GFP_KERNEL. kmalloc_node is never called. this is also wrong. i want to use kmalloc as first try for bigger sized. otherwise the kvmalloc fallback implementation makes no sense. so just change the if clause as i wrote it would be my suggestion. it has the correct behaviour

Reply:
> if kmalloc_node fails and size <= PAGE_SIZE must return NULL too

This is exactly what my implementation does... if kmalloc fails, `return ptr` is not called but `return NULL` is for !GFP_KERNEL.

> in your code if size is > PAGE_SIZE and flags are GFP_KERNEL. kmalloc_node is never called. this is also wrong. i want to use kmalloc as first try for bigger sized

Ouch, right. I will correct that

Reply:
no it does not. if kmalloc_node fails. and size <= page size return ptr is not called. this has nothing todo with the flags. any size < PAGE_SIZE which does not pass the kmalloc_node call, must return as failed and shall never use __vmalloc

Reply:
fixed